edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/metrics/MetricsCommandFactory.cs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Microsoft. All rights reserved.
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.InteropServices.ComTypes;
+using System.Threading;
+using System.Threading.Tasks;
+using Akka.Streams.Implementation;
+using App.Metrics.Counter;
+using App.Metrics.Formatters.Prometheus;
+using Microsoft.Azure.Devices.Edge.Util;
+using Microsoft.Azure.Devices.Edge.Util.Metrics;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Devices.Edge.Agent.Core
+{
+    public class MetricsCommandFactory : ICommandFactory
+    {
+        readonly ICommandFactory underlying;
+
+        public MetricsCommandFactory(ICommandFactory underlying)
+        {
+            this.underlying = Preconditions.CheckNotNull(underlying, nameof(underlying));
+        }
+
+        public async Task<ICommand> CreateAsync(IModuleWithIdentity module, IRuntimeInfo runtimeInfo)
+        {
+            FactoryMetrics.AddMessage(module.Module, FactoryMetrics.Command.Start);
+            return await this.underlying.CreateAsync(module, runtimeInfo);
+        }
+
+        public async Task<ICommand> UpdateAsync(IModule current, IModuleWithIdentity next, IRuntimeInfo runtimeInfo)
+        {
+            FactoryMetrics.AddMessage(current, FactoryMetrics.Command.Start);
+            FactoryMetrics.AddMessage(next.Module, FactoryMetrics.Command.Stop);
+            return await this.underlying.UpdateAsync(current, next, runtimeInfo);
+        }
+
+        public async Task<ICommand> UpdateEdgeAgentAsync(IModuleWithIdentity module, IRuntimeInfo runtimeInfo)
+        {
+            return await this.underlying.UpdateEdgeAgentAsync(module, runtimeInfo);
+        }
+
+        public async Task<ICommand> RemoveAsync(IModule module)
+        {
+            FactoryMetrics.AddMessage(module, FactoryMetrics.Command.Stop);
```

Comment:
why not report "remove" command?

Reply:
I easily can, I'm waiting to add more until meeting Thursday to discuss what metrics we want to collect.